curp/src/server/curp_node.rs

```diff
@@ -307,27 +307,36 @@ impl<C: 'static + Command> CurpNode<C> {
         let mut ticker = tokio::time::interval(curp.cfg().heartbeat_interval);
         ticker.set_missed_tick_behavior(MissedTickBehavior::Delay);
         let id = connect.id();
+        let batch_timeout = curp.cfg().batch_timeout;
 
         #[allow(clippy::integer_arithmetic, clippy::unwrap_used)]
         // tokio select internal triggered,
         loop {
             // grab the listener in the beginning to prevent missing sync events
-            let listener = sync_event.listen();
+            let listener: event_listener::EventListener = sync_event.listen();
 
             let Ok(sync_action) = curp.sync(id) else {
                 return;
             };
 
             match sync_action {
                 SyncAction::AppendEntries(ae) => {
-                    let result = Self::send_ae(connect.as_ref(), curp.as_ref(), ae).await;
-                    if let Err(err) = result {
-                        warn!("ae to {} failed, {err}", connect.id());
-                        if matches!(err, SendAEError::NotLeader) {
-                            return;
+                    // (hb_opt, entries) status combination
+                    // (false, empty) => send heartbeat to followers
+                    // (true, empty) => indicates that `batch_timeout` expired, and during this period there is not any log generated. Do nothing
+                    // (true | false, not empty) => send append entries
+                    if !hb_opt || !ae.entries.is_empty() {
+                        let result = Self::send_ae(connect.as_ref(), curp.as_ref(), ae).await;
+                        if let Err(err) = result {
+                            warn!("ae to {} failed, {err}", connect.id());
+                            if matches!(err, SendAEError::NotLeader) {
+                                return;
+                            }
+                        } else {
+                            hb_opt = true;
                         }
                     } else {
-                        hb_opt = true;
+                        hb_opt = false;
                     }
```

Comment:
And here hb_opt should not be set to false? If we skip an ae here, it doesn't mean that next ae should not be skipped. The hb_opt should only be set to false when heartbeat_timeout is reached?

Reply:
Fixed. The reason why I set `hb_opt` to false is that we had an if statement in the select block. And now it's gone, so we don't set it again.